cpp/cgi_progs/ini_validator.cc

```diff
@@ -0,0 +1,111 @@
+/** \file    ini_validator.cc
+ *  \brief   A CGI-tool to validate INI files
+ *  \author  Mario Trojan
+ */
+/*
+    Copyright (C) 2020, Tübingen University Library
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
+#include "FileUtil.h"
+#include "HtmlUtil.h"
+#include "IniFile.h"
+#include "StringUtil.h"
+#include "util.h"
+#include "WebUtil.h"
+
+
+namespace {
+
+
+std::string GetCGIParameterOrDefault(const std::multimap<std::string, std::string> &cgi_args,
+                                     const std::string &parameter_name,
+                                     const std::string &default_value = "")
+{
+    const auto key_and_value(cgi_args.find(parameter_name));
+    if (key_and_value == cgi_args.cend())
+        return default_value;
+
+    return key_and_value->second;
+}
+
+
+void ShowUploadForm() {
+    std::cout << "<h1>INI validator</h1>\n"
+              << "<p>Please paste the contents of the INI file here:</p>\n"
+              << "<form id=\"upload_form\" method=\"post\">\n"
+              << "\t<textarea name=\"ini_content\" rows=\"20\" cols=\"100\"></textarea>\n"
+              << "\t<input type=\"hidden\" name=\"action\" value=\"validate\">\n"
+              << "\t<br><br>\n"
+              << "\t<button onclick=\"document.getElementById(\"upload_form\").submit()\">Validate</button>\n"
+              << "</form>\n";
+}
+
+
+void Validate(const std::multimap<std::string, std::string> &cgi_args) {
+
+    std::cout << "<h1>Validate</h1>";
```

Comment:
I would also add a newline here because o/w when you do a "view source" in a browser, you get a huge mess.